File: app/src/main/java/com/sjn/stamp/controller/StampController.java
package com.sjn.stamp.controller;

import android.content.Context;
import android.support.v4.media.MediaMetadataCompat;

import com.sjn.stamp.constant.CategoryType;
import com.sjn.stamp.db.CategoryStamp;
import com.sjn.stamp.db.Song;
import com.sjn.stamp.db.SongStamp;
import com.sjn.stamp.db.dao.CategoryStampDao;
import com.sjn.stamp.db.dao.SongStampDao;
import com.sjn.stamp.utils.LogHelper;
import com.sjn.stamp.utils.RealmHelper;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import io.realm.Realm;

public class StampController {
    @SuppressWarnings("unused")
    private static final String TAG = LogHelper.makeLogTag(StampController.class);
    @SuppressWarnings({"FieldCanBeLocal", "unused"})
    private Context mContext;
    private List<Listener> mListenerSet = new ArrayList<>();

    public void addListener(Listener listener) {
        mListenerSet.add(listener);
    }

    public void removeListener(Listener listener) {
        mListenerSet.remove(listener);
    }

    public void remove(String stamp) {
        Realm realm = RealmHelper.getRealmInstance();
        SongStampDao.INSTANCE.remove(realm, stamp);
        CategoryStampDao.INSTANCE.remove(realm, stamp);
        realm.close();
        notifyStampChange();
    }

    public interface Listener {
        void onStampChange();
    }

    void notifyStampChange() {
        for (Listener listener : mListenerSet) {
            listener.onStampChange();
        }
    }

    public StampController(Context context) {
        mContext = context;
    }

    public List<String> findAll() {
        List<String> stampList = new ArrayList<>();
        Realm realm = RealmHelper.getRealmInstance();
        for (SongStamp songStamp : SongStampDao.INSTANCE.findAll(realm)) {
            stampList.add(songStamp.getName());
        }
        realm.close();
        return stampList;
    }

    public boolean register(String name) {
        Realm realm = RealmHelper.getRealmInstance();
        boolean result = SongStampDao.INSTANCE.save(realm, name);
        realm.close();
        notifyStampChange();
        return result;
    }

    public Map<String, List<MediaMetadataCompat>> getAllSongList(final Map<String, MediaMetadataCompat> musicListById) {
        Map<String, Map<String, MediaMetadataCompat>> songStampMap = new ConcurrentHashMap<>();
        Realm realm = RealmHelper.getRealmInstance();
        for (SongStamp songStamp : SongStampDao.INSTANCE.findAll(realm)) {
            put(songStampMap, songStamp.getName(), createTrackMap(songStamp));
        }
        Map<String, Map<CategoryType, List<String>>> stampQueryMap = new ConcurrentHashMap<>();
        for (CategoryStamp categoryStamp : CategoryStampDao.INSTANCE.findAll(realm)) {
            put(stampQueryMap, categoryStamp);
        }
        realm.close();

        Map<String, Map<String, MediaMetadataCompat>> categoryStampMap = searchMusic(musicListById, stampQueryMap);
        merge(songStampMap, categoryStampMap);

        Map<String, List<MediaMetadataCompat>> stampSongMap = new ConcurrentHashMap<>();
        for (Map.Entry<String, Map<String, MediaMetadataCompat>> entry : songStampMap.entrySet()) {
            stampSongMap.put(entry.getKey(), new ArrayList<>(entry.getValue().values()));
        }
        return stampSongMap;
    }

    private void merge(Map<String, Map<String, MediaMetadataCompat>> songStampMap, Map<String, Map<String, MediaMetadataCompat>> categoryStampMap) {
        for (String stamp : categoryStampMap.keySet()) {
            if (songStampMap.containsKey(stamp)) {
                songStampMap.get(stamp).putAll(categoryStampMap.get(stamp));
            } else {
                songStampMap.put(stamp, categoryStampMap.get(stamp));
            }
        }
    }

    private Map<String, MediaMetadataCompat> createTrackMap(SongStamp songStamp) {
        Map<String, MediaMetadataCompat> trackMap = new ConcurrentHashMap<>();
        if (songStamp.getSongList() == null) {
            return trackMap;
        }
        for (Song song : songStamp.getSongList()) {
            trackMap.put(song.getMediaId(), song.buildMediaMetadataCompat());
        }
        return trackMap;
    }

    private Map<String, Map<String, MediaMetadataCompat>> searchMusic(final Map<String, MediaMetadataCompat> musicListById, Map<String, Map<CategoryType, List<String>>> queryMap) {
        Map<String, Map<String, MediaMetadataCompat>> result = new ConcurrentHashMap<>();
        if (musicListById == null || queryMap == null) {
            return result;
        }
        for (MediaMetadataCompat track : musicListById.values()) {
            for (Map.Entry<String, Map<CategoryType, List<String>>> entry1 : queryMap.entrySet()) {
                for (Map.Entry<CategoryType, List<String>> entry : entry1.getValue().entrySet()) {
                    if (entry.getValue().contains(track.getString(entry.getKey().getKey()).toLowerCase(Locale.getDefault()))) {
                        put(result, entry1.getKey(), track);
                    }
                }
            }
        }
        return result;
    }

    private void put(Map<String, Map<String, MediaMetadataCompat>> stampMap, String stampName, MediaMetadataCompat track) {
        if (stampMap == null) {
            return;
        }
        if (stampMap.containsKey(stampName) && !stampMap.get(stampName).isEmpty()) {
            stampMap.get(stampName).put(track.getDescription().getMediaId(), track);
        } else {
            Map<String, MediaMetadataCompat> trackMap = new ConcurrentHashMap<>();
            trackMap.put(track.getDescription().getMediaId(), track);
            stampMap.put(stampName, trackMap);
        }
    }

    private void put(Map<String, Map<String, MediaMetadataCompat>> stampMap, String stampName, Map<String, MediaMetadataCompat> trackMap) {
        if (stampMap == null || trackMap == null) {
            return;
        }
        if (stampMap.containsKey(stampName)) {
            stampMap.get(stampName).putAll(trackMap);
        } else {
            stampMap.put(stampName, trackMap);
        }
    }

    private void put(Map<String, Map<CategoryType, List<String>>> stampQueryMap, CategoryStamp categoryStamp) {
        if (stampQueryMap == null || categoryStamp == null) {
            return;
        }
        if (stampQueryMap.containsKey(categoryStamp.getName())) {
            putQuery(stampQueryMap.get(categoryStamp.getName()), categoryStamp);
        } else {
            Map<CategoryType, List<String>> queryMap = new ConcurrentHashMap<>();
            putQuery(queryMap, categoryStamp);
            stampQueryMap.put(categoryStamp.getName(), queryMap);
        }
    }

    private void putQuery(Map<CategoryType, List<String>> queryMap, CategoryStamp categoryStamp) {
        if (queryMap == null || categoryStamp == null) {
            return;
        }
        CategoryType categoryType = CategoryType.Companion.of(categoryStamp.getType());
        if (categoryType == null || categoryStamp.getValue() == null) {
            return;
        }
        String query = categoryStamp.getValue().toLowerCase(Locale.getDefault());
        if (queryMap.containsKey(categoryType) && !queryMap.get(categoryType).isEmpty()) {
            queryMap.get(categoryType).add(query);
        } else {
            queryMap.put(categoryType, new ArrayList<>(Collections.singletonList(query)));
        }
    }
}

File: app/src/main/java/com/sjn/stamp/controller/StampController.kt
package com.sjn.stamp.controller

import android.content.Context
import android.support.v4.media.MediaMetadataCompat
import com.sjn.stamp.model.CategoryStamp
import com.sjn.stamp.model.SongStamp
import com.sjn.stamp.model.constant.CategoryType
import com.sjn.stamp.model.dao.CategoryStampDao
import com.sjn.stamp.model.dao.SongDao
import com.sjn.stamp.model.dao.SongStampDao
import com.sjn.stamp.utils.AnalyticsHelper
import com.sjn.stamp.utils.LogHelper
import com.sjn.stamp.utils.RealmHelper
import io.realm.Realm
import java.util.*
import java.util.concurrent.ConcurrentHashMap

class StampController(private val context: Context) {
    private val listenerSet = ArrayList<Listener>()

    interface Listener {
        fun onStampChange()
    }

    val isMyStampExists: Boolean
        get() =
            RealmHelper.realmInstance.use { realm ->
                SongStampDao.isMyStampExists(realm)
            }

    val isSmartStampExists: Boolean
        get() =
            RealmHelper.realmInstance.use { realm ->
                SongStampDao.isSmartStampExists(realm)
            }

    fun addListener(listener: Listener) {
        listenerSet.add(listener)
    }

    fun removeListener(listener: Listener) {
        listenerSet.remove(listener)
    }

    fun register(name: String, isSystem: Boolean): Boolean {
        if (name.isEmpty()) {
            return false
        }
        if (RealmHelper.realmInstance.use { SongStampDao.find(it, name, isSystem) } != null) {
            return false
        }
        RealmHelper.realmInstance.use { realm -> SongStampDao.findOrCreate(realm, name, isSystem) }
        notifyStampChange()
        AnalyticsHelper.trackStamp(context, name)
        return true
    }

    fun isCategoryStamp(name: String, isSystem: Boolean, mediaId: String): Boolean {
        val mediaMetadata = SongController(context).resolveMediaMetadata(mediaId) ?: return false
        RealmHelper.realmInstance.use { realm ->
            return CategoryStampDao.findByName(realm, name, isSystem).find { categoryStamp -> categoryStamp.contain(mediaMetadata) } != null
        }
    }

    fun removeSong(name: String, isSystem: Boolean, mediaId: String): Boolean {
        val mediaMetadata = SongController(context).resolveMediaMetadata(mediaId) ?: return false
        RealmHelper.realmInstance.use { realm ->
            val song = SongDao.findByMediaMetadata(realm, mediaMetadata) ?: return false
            return SongStampDao.remove(realm, song.id, name, isSystem)
        }
    }

    fun delete(stamp: String, isSystem: Boolean) {
        RealmHelper.realmInstance.use { realm ->
            delete(realm, stamp, isSystem)
        }
        notifyStampChange()
    }

    fun delete(realm: Realm, stamp: String, isSystem: Boolean) {
        SongStampDao.delete(realm, stamp, isSystem)
        CategoryStampDao.delete(realm, stamp, isSystem)
    }

    fun findAll(): List<String> =
            RealmHelper.realmInstance.use { realm ->
                SongStampDao.findAll(realm).map { it.name }
            }

    fun findAllMyStamps(): List<String> {
        return RealmHelper.realmInstance.use { realm ->
            SongStampDao.findAll(realm, false).map { it.name }
        }
    }

    fun createStampMap(musicListById: MutableMap<String, MediaMetadataCompat>, stampToList: MutableMap<String, MutableList<MediaMetadataCompat>>, isSystem: Boolean): MutableMap<String, MutableList<MediaMetadataCompat>> {
        val stampToMap = findSongStamp(isSystem)
        if (!isSystem) {
            stampToMap.merge(findCategoryStamp(musicListById))
        }
        for ((key, value) in stampToMap) {
            stampToList[key] = ArrayList(value.values)
        }
        return stampToList
    }

    internal fun notifyStampChange() {
        for (listener in listenerSet) {
            listener.onStampChange()
        }
    }

    private fun findSongStamp(isSystem: Boolean): MutableMap<String, MutableMap<String, MediaMetadataCompat>> {
        val songStampMap = ConcurrentHashMap<String, MutableMap<String, MediaMetadataCompat>>()
        RealmHelper.realmInstance.use { realm ->
            for (songStamp in SongStampDao.findAll(realm, isSystem)) {
                songStampMap.putSongMap(songStamp.name, songStamp.songMap())
            }
        }
        return songStampMap
    }

    private fun findCategoryStamp(musicListById: MutableMap<String, MediaMetadataCompat>): MutableMap<String, MutableMap<String, MediaMetadataCompat>> {
        val stampQueryMap = ConcurrentHashMap<String, MutableMap<CategoryType, MutableList<String>>>()
        RealmHelper.realmInstance.use { realm ->
            for (categoryStamp in CategoryStampDao.findAll(realm)) {
                stampQueryMap.putCategoryStamp(categoryStamp)
            }
        }
        return stampQueryMap.executeQuery(musicListById)
    }

    private fun MutableMap<String, MutableMap<String, MediaMetadataCompat>>.merge(newMap: MutableMap<String, MutableMap<String, MediaMetadataCompat>>): MutableMap<String, MutableMap<String, MediaMetadataCompat>> {
        for ((key, value) in newMap) {
            if (containsKey(key)) {
                this[key]?.putAll(value)
            } else {
                put(key, value)
            }
        }
        return this
    }

    /*
        'favorite' ->  {'mediaId_1' -> song1, 'mediaId_2' -> song2}
        'recent'   ->  {'mediaId_1' -> song1, 'mediaId_2' -> song2}
     */
    private fun MutableMap<String, MutableMap<CategoryType, MutableList<String>>>.executeQuery(musicListById: MutableMap<String, MediaMetadataCompat>): MutableMap<String, MutableMap<String, MediaMetadataCompat>> {
        val result = ConcurrentHashMap<String, MutableMap<String, MediaMetadataCompat>>()
        for (song in musicListById.values) {
            for ((stampName, categoryQueryMap) in this) {
                for ((categoryType, queryList) in categoryQueryMap) {
                    if (queryList.contains(song.getString(categoryType.key).toLowerCase(Locale.getDefault()))) {
                        result.putSong(stampName, song)
                    }
                }
            }
        }
        return result
    }

    /*
        'favorite' ->  {'mediaId_1' -> song1, 'mediaId_2' -> song2}
        'recent'   ->  {'mediaId_1' -> song1, 'mediaId_2' -> song2}
     */
    private fun MutableMap<String, MutableMap<String, MediaMetadataCompat>>.putSong(stampName: String, song: MediaMetadataCompat) {
        if (this.containsKey(stampName) && !this[stampName]!!.isEmpty()) {
            this[stampName]?.put(song.description.mediaId!!, song)
        } else {
            val songMap = ConcurrentHashMap<String, MediaMetadataCompat>()
            songMap[song.description.mediaId!!] = song
            this[stampName] = songMap
        }
    }

    /*
        'favorite' ->  {'mediaId_1' -> song1, 'mediaId_2' -> song2}
        'recent'   ->  {'mediaId_1' -> song1, 'mediaId_2' -> song2}
     */
    private fun MutableMap<String, MutableMap<String, MediaMetadataCompat>>.putSongMap(stampName: String, songMap: MutableMap<String, MediaMetadataCompat>) {
        if (this.containsKey(stampName)) {
            this[stampName]?.putAll(songMap)
        } else {
            this[stampName] = songMap
        }
    }

    /*
        'favorite' ->  {CategoryType.ALBUM -> ['title1', 'title2'], CategoryType.ARTIST -> ['artist1', 'artist2']}
        'recent'   ->  {CategoryType.ALBUM -> ['title3', 'title4'], CategoryType.ARTIST -> ['artist5', 'artist6']}
     */
    private fun MutableMap<String, MutableMap<CategoryType, MutableList<String>>>.putCategoryStamp(categoryStamp: CategoryStamp) {
        if (this.containsKey(categoryStamp.name)) {
            this[categoryStamp.name]?.putQuery(categoryStamp)
        } else {
            val queryMap = ConcurrentHashMap<CategoryType, MutableList<String>>()
            queryMap.putQuery(categoryStamp)
            this[categoryStamp.name] = queryMap
        }
    }

    /*
        CategoryType.ALBUM  -> ['title1', 'title2']
        CategoryType.ARTIST -> ['artist1', 'artist2']
     */
    private fun MutableMap<CategoryType, MutableList<String>>.putQuery(categoryStamp: CategoryStamp) {
        val categoryType = CategoryType.of(categoryStamp.type) ?: return
        val query = categoryStamp.value.toLowerCase(Locale.getDefault())
        if (this.containsKey(categoryType) && !this[categoryType]!!.isEmpty()) {
            this[categoryType]?.add(query)
        } else {
            this[categoryType] = ArrayList(listOf(query))
        }
    }

    private fun SongStamp.songMap(): MutableMap<String, MediaMetadataCompat> {
        val songMap = ConcurrentHashMap<String, MediaMetadataCompat>()
        for (song in this.songList) {
            songMap[song.mediaId] = song.buildMediaMetadataCompat()
        }
        return songMap
    }

    companion object {
        @Suppress("unused")
        private val TAG = LogHelper.makeLogTag(StampController::class.java)
    }
}
